File: okreplay-core/src/main/java/okreplay/Absent.java
package okreplay;

import java.util.Collections;
import java.util.Set;
import java.util.function.Function;

import javax.annotation.Nullable;

import static okreplay.Util.checkNotNull;

final class Absent<T> extends Optional<T> {
  private static final Absent<Object> INSTANCE = new Absent<>();

  @SuppressWarnings("unchecked") // implementation is "fully variant"
  static <T> Optional<T> withType() {
    return (Optional<T>) INSTANCE;
  }

  private Absent() {}

  @Override
  boolean isPresent() {
    return false;
  }

  @Override
  T get() {
    throw new IllegalStateException("Optional.get() cannot be called on an absent value");
  }

  @Override
  T or(T defaultValue) {
    return checkNotNull(defaultValue, "use Optional.orNull() instead of Optional.or(null)");
  }

  @SuppressWarnings("unchecked") // safe covariant cast
  @Override Optional<T> or(Optional<? extends T> secondChoice) {
    return (Optional<T>) checkNotNull(secondChoice);
  }

  @Override @Nullable T orNull() { return null;
  }

  @Override
  Set<T> asSet() {
    return Collections.emptySet();
  }

  @Override
  <V> Optional<V> transform(Function<? super T, V> function) {
    checkNotNull(function);
    return Optional.absent();
  }

  @Override public boolean equals(@Nullable Object object) {
    return object == this;
  }

  @Override public int hashCode() {
    return 0x79a31aac;
  }

  @Override public String toString() {
    return "Optional.absent()";
  }

  private Object readResolve() {
    return INSTANCE;
  }

  private static final long serialVersionUID = 0;
}

File: okreplay-core/src/main/java/okreplay/Absent.kt
package okreplay

import okreplay.Util.checkNotNull
import java.util.function.Function

internal class Absent<T> private constructor() : Optional<T>() {

  override val isPresent: Boolean = false

  override fun get(): T {
    throw IllegalStateException("Optional.get() cannot be called on an absent value")
  }

  override fun or(defaultValue: T): T {
    return checkNotNull(defaultValue, "use Optional.orNull() instead of Optional.or(null)")
  }

  @Suppress("UNCHECKED_CAST") // safe covariant cast
  override fun or(secondChoice: Optional<out T>): Optional<T> {
    return checkNotNull(secondChoice) as Optional<T>
  }

  override fun orNull(): T? {
    return null
  }

  override fun asSet(): Set<T> {
    return emptySet()
  }

  override fun <V> transform(function: Function<in T, V>): Optional<V> {
    checkNotNull(function)
    return Optional.absent()
  }

  override fun equals(`object`: Any?): Boolean {
    return `object` === this
  }

  override fun hashCode(): Int {
    return 0x79a31aac
  }

  override fun toString(): String {
    return "Optional.absent()"
  }

  private fun readResolve(): Any {
    return INSTANCE
  }

  companion object {
    private val INSTANCE = Absent<Any>()

    @Suppress("UNCHECKED_CAST") // implementation is "fully variant"
    fun <T> withType(): Optional<T> {
      return INSTANCE as Optional<T>
    }

    private val serialVersionUID: Long = 0
  }
}